docs/design-documents/platform/Statistics/NetworkStatistics.md

```diff
@@ -0,0 +1,186 @@
+                                                                |
+                            -----------------------------------------------------------------------------------
+                           |                            |                           |                         |
+                           |                            |                           |                         |
+,----------------------------------------.      ,------------------------.      ,--------------------.      ,----.
+|                 <Module_name>Stats     |      |   EMACInterfaceStats   |      |   LWIPStackStats   |  ... |    |
+|----------------------------------------|      `------------------------'      `--------------------'      `----'
+|+ <module>_stats_t _stats;              |                 |Uses                          |Uses
+|----------------------------------------|                 |                              |
+|+ inline void log_xx();                 |                 |                              |
+|  ...                                   |                 |                              |
+|+ virtual void read_stats(...);         |                 |                              |
+`----------------------------------------'                 |                              |
+               |Uses                                       |                              |
+               |                                           |                              |
+,---------------------------------.     ,---------------------------------.   ,-----------------------------.
+| Class: <Module>                 |     |Class: EMACInterface             |   |Class: LWIP                  |
+|#Object:<Module_name>Stats stats;|     |#Object:EMACInterfaceStats stats;|   |#Object:LWIPStackStats stats;|
+`---------------------------------'     `---------------------------------'   `-----------------------------'
+```
+
+# Detailed design
+
+## StatsList Class
+An abstract class used to create a linked list entry of every module reporting statistics data. Static API's in StatsList class can be used to read statistics data stored in linked list.
+
+#### Data Structures
+
+1. Statistic types for segregation of different modules and can also be used for filtering statistics data.
+```
+/** Stats type */
+typedef enum {
+    STATS_RTOS_GEN = 0x0,
+
+    STATS_PLATFORM_GEN = 0x20,
+
+    STATS_STORAGE_GEN = 0x40,
+
+    STATS_NW_GEN = 0x60,
+    STATS_NW_INTERFACE_EMAC,
+    STATS_NW_INTERFACE_WIFI,
+    STATS_NW_INTERFACE_MESH,
+    STATS_NW_INTERFACE_CELLULAR,
+
+    STATS_NW_STACK_LWIP,
+    STATS_NW_STACK_WIFI,
+}mbed_stats_type_t;
+```
+1. Statistic Info Structure - Structure used to get statistics data as binary blob from the StatsList class.
+
+```
+typedef struct  {
+    mbed_stats_type_t stats_type;
+    uint32_t buf_size;
+    void *buf;
+}stats_info_t;
+```
+
+#### API's
+
+1. static int get_each(stats_info_t *stats, int count);
```

Comment:
By fetch, do you mean that the data is going to be copied into the `stats_info_t *stats` and.. linked by how.. This is very confusing API. How much do we need to allocate memory to get the statistics and can it change in the runtime?

Could we just return the link to the head of the linked list that we already have?

Reply:
It is implemented the same way as other stats, user need to allocate memory for array of `stats_info_t` structure based on the `count` requested

Reply:
User is not aware of linked list, API's report stats as an array to structure. Memory allocated by user will be equivalent to `size of (stats_info_t ) * count` . Users do not need to allocate memory for RAM buffers of network interface and stack stats